File: Leetcode-one-by-one/src/The k-th Lexicographical String of All Happy Strings of Length n.swift
class Solution {
    func getHappyString(_ n: Int, _ k: Int) -> String {
        var k = k
        let total = 3 * (1 << (n - 1))
        if k > total { return "" }

        var res = ""
        var prev: Character = "#"

        for i in 0..<n {
            for c in ["a", "b", "c"] {
                if c.first! == prev { continue }

                let remain = n - i - 1
                let count = 1 << remain   // 2^(remaining)

                if k > count {
                    k -= count
                } else {
                    res.append(c)
                    prev = c.first!
                    break
                }
            }
        }

        return res
    }
}

File: Leetcode-one-by-one/src/The k-th Lexicographical String of All Happy Strings of Length n.java
class Solution {
    public String getHappyString(int n, int k) {
        int total = 3 * (1 << (n - 1));
        if (k > total) return "";

        StringBuilder res = new StringBuilder();
        char prev = '#';

        for (int i = 0; i < n; i++) {
            for (char c : new char[]{'a','b','c'}) {

                if (c == prev) continue;

                int remain = n - i - 1;
                int count = 1 << remain;   // 2^(remaining)

                if (k > count) {
                    k -= count;
                } else {
                    res.append(c);
                    prev = c;
                    break;
                }
            }
        }

        return res.toString();
    }
}
